File: code/750/solution.py
# -*- coding: UTF-8 -*-
from random import random
from math import *

num = [0] * 1000
dis = [[0 for i in range(1000)] for j in range(1000)]
ans = [[0 for i in range(1000)] for j in range(1000)]


def G(n):
    now = 1
    for i in range(1, n + 1):
        now = now * 3 % (n + 1)
        num[i] = now

    for i in range(1, n + 1):
        for j in range(1, n + 1):
            dis[num[i]][num[j]] = abs(i - j)

    for j in range(1, n + 1):
        for i in range(j, 0, -1):
            if i == j:
                ans[i][j] = 0
                continue
            ans[i][j] = 10 ** 9
            for k in range(i, j):
                ans[i][j] = min(ans[i][j], ans[i][k] + ans[k + 1][j] + dis[k][j])
    return ans[1][n]


if __name__ == "__main__":
    print(G(6))
    print(G(16))
    print(G(976))

File: code/750/solution.cpp
#include <iostream>/*{{{*/
#include <math.h>
using namespace std; /*}}}*/
int num[1000];
int dis[1000][1000];
int ans[1000][1000];
int G(int n) {
    int now = 1;
    for (int i = 1; i <= n; i++) {
        now    = now * 3 % (n + 1);
        num[i] = now;
    }
    for (int i = 1; i <= n; i++) {
        for (int j = 1; j <= n; j++) {
            dis[num[i]][num[j]] = abs(i - j);
        }
    }

    for (int j = 1; j <= n; j++) {
        for (int i = j; i >= 1; i--) {
            if (i == j) {
                ans[i][j] = 0;
                continue;
            }
            ans[i][j] = 1e9;
            for (int k = i; k < j; k++) {
                ans[i][j] = min(ans[i][j], ans[i][k] + ans[k + 1][j] + dis[k][j]);
            }
        }
    }
    return ans[1][n];
}
int main() {
    cout << G(6) << endl;
    cout << G(16) << endl;
    cout << G(976) << endl;
    return 0;
}
